### ISSUE TITLE
jlink does not produce a working runtime, VM is missing

### ISSUE THREAD

**Issue by jdekonin** [ibmruntimes/openj9-openjdk-jdk9#21]
From an OpenJ9 build, run jlink to create a new runtime.

i.e. The following produces a new runtime in the "base" directory, containing only the java.base module.
bin/jlink --module-path jmods --add-modules java.base --output base

Try to run the new runtime, it doesn't work because the J9 VM is missing.
base/bin/java
Error: missing j9vm' JVM at/bluebird/builds/bld_361845/sdk/xa6490_openj9/jdk/base/lib/j9vm/libjvm.so'.
Please install or use the JRE or JDK that contains these missing components.

At least the following is missing. These should all be included in the java.base.jmod
lib/J9TraceFormat.dat
lib/OMRTraceFormat.dat
lib/compressedrefs (the entire directory - this is the VM proper)
lib/j9vm/libjvm.so
lib/java*.properties
lib/options.default

**Comment by andrew-m-leonard** [comment]
@jdekonin I'm discovering I can't accurately fix this, I think this needs a J9VM expert who knows what lib to copy where to take a look...

**Comment by andrew-m-leonard** [comment]
This is what I have tried so far in closed/OpenJ9.gmk build-j9 recipe:
        # Copy J9 VM support modules
        @$(MKDIR) -p $(OUTPUT_ROOT)/support/modules_libs/java.base/j9vm
        @$(MKDIR) -p $(OUTPUT_ROOT)/support/modules_libs/java.base/compressedrefs
        @$(CP) -p $(OUTPUT_ROOT)/vm/$(LIBRARY_PREFIX)jvm$(SHARED_LIBRARY_SUFFIX) $(OUTPUT_ROOT)/support/modules_libs/java.base/j9vm
        @$(CP) -p $(OUTPUT_ROOT)/vm/$(LIBRARY_PREFIX)*$(SHARED_LIBRARY_SUFFIX) $(OUTPUT_ROOT)/support/modules_libs/java.base/compressedrefs
        @$(CP) -p $(OUTPUT_ROOT)/vm/java*.properties $(OUTPUT_ROOT)/support/modules_libs/java.base
        @$(CP) -p $(OUTPUT_ROOT)/vm/J9TraceFormat.dat $(OUTPUT_ROOT)/support/modules_libs/java.base
        @$(CP) -p $(OUTPUT_ROOT)/vm/OMRTraceFormat.dat $(OUTPUT_ROOT)/support/modules_libs/java.base
        @$(CP) -p $(OUTPUT_ROOT)/vm/options.default $(OUTPUT_ROOT)/support/modules_libs/java.base

**Comment by pshipton** [comment]
You may also need to copy j9vm/libjvm.so to modules_libs/java.base/server

**Comment by pshipton** [comment]
Although java starts without having the lib/server/libjvm.so, but this would make a jlinked version the same as the jdk build

**Comment by pshipton** [comment]
Many libs in lib/compressedrefs need to be added (see *), although they should remain in the compressedrefs subfolder of java.base. What you are trying to (re)create is the same directory structure as found in the OpenJ9 build itself. I suggest comparing the output from jlink to the OpenJ9 build and checking for differences, and then fix the script accordingly. It is expected the output from jlink to have fewer files since only java.base is included, but the entire VM needs to be included. Andrew's list of VM files looks correct. Many lib/lib*.so files will be missing as expected. 

One thing to watch out for, there are two different copies of libjvm.so. The one in compressedrefs is bigger, and the one in j9vm is smaller. The copy in server should be the same as the j9vm one. I think the OpenJ9 build is incorrect, as the server/libjvm.so is the same as the compressedrefs one.

* lib/compressedrefs contains a number of libs only needed for testing, these ones don't need to be added to java.base. Here is a list of the necessary compressedrefs libs:
libcuda4j29.so
libffi29.so
libhyprtshim29.so
libhythr.so
libj9dmp29.so
libj9gc29.so
libj9gcchk29.so
libj9hookable29.so
libj9jit29.so
libj9jnichk29.so
libj9jvmti29.so
libj9prt29.so
libj9shr29.so    -> put in openj9.sharedclasses
libj9thr29.so
libj9trc29.so
libj9vm29.so
libj9vmchk29.so
libj9vrb29.so
libj9zlib29.so
libjclse9_29.so
libjsig.so
libjvm.so
libmanagement.so          -> put in java.management
libmanagement_ext.so    -> put in java.management
libomrsig.so

**Comment by jdekonin** [comment]
There are macros in place to perform these task, the $(CP) shouldn't be used in this way.  This should also be figured out from the openjdk extensions side (??).  During the compile there are lines like "Creating support/modules_libs/java.base/libnio.so from 24 file(s)"  We need to extend this .gmk functionality to include openj9 content on a module basis.

**Comment by yongja79** [comment]
@jdekonin Are you saying we want to recompile the files? The lines like "Creating support/modules_libs/java.base/libnio.so from 24 file(s)" are happening in the NativeCompilation.gmk. As the files we want already exist in the vm directory why isn't copying them across enough? We already do this for several other files...

**Comment by andrew-m-leonard** [comment]
@yongja79 Looks good Jason :-)

**Comment by andrew-m-leonard** [comment]
admin@b6c926402e69:~/ccm/openj9Public/build/linux-x86_64-normal-server-release/images/jdk$ base/bin/java -version
openjdk version "9-internal"
OpenJDK Runtime Environment (build 9-internal+0-adhoc..openj9Public)
Eclipse OpenJ9 VM (build 2.9, JRE 9 Linux amd64-64 Compressed References 20171017_000000 (JIT enabled, AOT enabled)
OpenJ9   - a82451c
OMR      - e705c80
OpenJDK  - 20e6fd5 based on jdk-9+181)

**Comment by keithc-ca** [comment]
This isn't quite fixed yet. With all repos up to date, `make clean && make`fails:
```
Error: missing `j9vm' JVM at `/home/openj9/openj9-openjdk-jdk9/build/linux-x86_64-normal-server-release/jdk/lib/j9vm/libjvm.so'.
Please install or use the JRE or JDK that contains these missing components.
make[3]: *** [/home/openj9/openj9-openjdk-jdk9/build/linux-x86_64-normal-server-release/jdk/_packages_attribute.done] Error 4
make[2]: *** [exploded-image-optimize] Error 1
```

**Comment by keithc-ca** [comment]
Removed my comment about #34 not being fixed yet (which doesn't belong here).

**Comment by pshipton** [comment]
Tried a few things myself, looks good. I assume this can be closed now.

BTW, the pull request would have automatically closed this if it contained "Fixes #21" instead of "Fixes issue #21".
